sockeye/training.py

```diff
@@ -328,9 +337,28 @@ def _determine_improvement(self, val_metrics: List[loss.LossMetric]) -> bool:
         for val_metric in val_metrics:
             if val_metric.name == self.config.early_stopping_metric:
                 value = val_metric.get()
-                if utils.metric_value_is_better(value,
-                                                self.state.best_metric,
-                                                self.config.early_stopping_metric):
+                # When using Horovod, the primary worker makes an authoritative
+                # check of whether metric value has improved and broadcasts the
+                # result to secondary workers.  Non-determinism in the order of
+                # GPU operations can lead to slight numeric variation across
+                # workers, causing potential desync if each worker makes its own
+                # check for key training decisions (reducing learning rate,
+                # early stopping, etc.).
+                if self.config.using_horovod and horovod_mpi.hvd.rank() > 0:
```

Comment:
For code clarity, would it make sense to define these checks in a separate method, something along the lines of `self.isMaster()` and `self.isWorker()`?

Reply:
That's a good idea.  We can add those convenience checks to the `horovod_mpi` module.

Reply:
I found another way to refactor this.  Adding a dedicated `using_horovod()` function to the `horovod_mpi` module means we don't have to pass around `using_horovod` values and can always check directly with the source (all things related to Horovod or MPI go through that optional module).

Taking another look, I do like explicitly making two checks: whether Horovod is running and what type of worker the current process is.

What do you think?